src/mumble/AudioInput.cpp

```diff
@@ -27,6 +27,102 @@ extern "C" {
 }
 #endif
 
+void Resynchronizer::addMic(short *mic)
+{
+	bool drop = false;
+	{
+		std::unique_lock<std::mutex> l(m);
+		micQueue.push_back(mic);
+		switch(state)
+		{
+			case S0:  state = S1a; break;
+			case S1a: state = S2;  break;
+			case S1b: state = S2;  break;
+			case S2:  state = S3;  break;
+			case S3:  state = S4a; break;
+			case S4a: state = S5;  break;
+			case S4b: drop = true; break;            
+			case S5:  drop = true; break;
+		}
+		if(drop)
+		{
+			delete[] micQueue.front();
+			micQueue.pop_front();
+		}
+	}
+	if(bDebugPrintQueue)
+	{
+		if(drop) printf("dropped microphone chunk due to overflow\n");
+		printQueue('+');
+	}
+}
+
+AudioChunk Resynchronizer::addSpeaker(short *speaker)
+{
+	AudioChunk result;
+	bool drop = false;
+	{
+		std::unique_lock<std::mutex> l(m);
+		switch(state)
+		{
+			case S0:  drop = true; break;
+			case S1a: drop = true; break;
+			case S1b: state = S0;  break;
+			case S2:  state = S1b; break;
+			case S3:  state = S2;  break;
+			case S4a: state = S3;  break;
+			case S4b: state = S3;  break;
+			case S5:  state = S4b; break;
+		}
+		if(drop == false)
+		{
+			result = AudioChunk(micQueue.front(), speaker);
+			micQueue.pop_front();
+		}
+	}
+	if(drop) delete[] speaker;
+	if(bDebugPrintQueue)
+	{
+		if(drop) printf("dropped speaker chunk due to underflow\n");
```

Comment:
For the sake of consistency I think this should be `qWarning` instead of `printf` as we're using Qt's logging stuff everywhere else as well. With that you'll also not have to add the NL at the end explicitly :)